chain-impl-mockchain/src/ledger/ledger.rs

```diff
@@ -997,11 +995,20 @@ impl Ledger {
                     new_ledger.apply_transaction(&fragment_id, &tx, block_date, ledger_params)?;
 
                 // we've just verified that this is a valid transaction (i.e. contains 1 input and 1 witness)
-                let account_id = match tx.inputs().iter().next().unwrap().to_enum() {
-                    InputEnum::UtxoInput(_) => {
-                        return Err(Error::OwnerStakeDelegationInvalidTransaction);
+                let account_id = match tx
+                    .inputs()
+                    .iter()
+                    .map(|input| input.to_enum())
+                    .zip(tx.witnesses().iter())
+                    .next()
+                    .unwrap()
+                {
+                    (InputEnum::AccountInput(account_id, _), Witness::Account(_, _)) => account_id
```

Comment:
why do we care about the witness here? I thought it was already verified in the `apply_transaction` call before

Reply:
Well, I changed the ledger to only allow voting with normal accounts (and not multisig accounts), because we were not taking the stake of those into account before, and we're still not doing it now.

But account inputs are the same for both, it's the witness type the thing that discriminates which kind of account is being used.

Reply:
It feels like there's a bit of a mismatch in the domain representation if we have to look at the witness to differentiate account types that comes with different features.
Like, the witness should really just be the spending credential for whatever is the input, and not be used for anything else or carry this additional information IMHO.

Reply:
I looks to me like the only type to appear in public interfaces should be `UnspecifiedAccountIdentifier` or `AccountIdentifier`, as there's no particular reason to prefer any of the two in term of functionalities (except for missing implementation, but at that point it should just return an error).

Reply:
>  Like, the witness should really just be the spending credential for whatever is the input, and not be used for anything else or carry this additional information IMHO.

It seems like an optimization to save space in the accounts (so they can be 32 bytes), but I don't know what was Vincent thinking. Multisig accounts seem like a half-baked feature overall.

> I looks to me like the only type to appear in public interfaces should be UnspecifiedAccountIdentifier or AccountIdentifier, as there's no particular reason to prefer any of the two in term of functionalities (except for missing implementation, but at that point it should just return an error).

I changed things to `account::Identifier` because I think it's clearer when reading the code that it only works with normal accounts. When I found the tests using `UnspecifiedAccountIdentifier` I had to go and check if I had to include the multisig ledger or not, only to realize that it we were ignoring it.

And currently `MintToken` uses `account::Identifier` too, so I think it's more consistent overall.

Reply:
> And currently `MintToken` uses `account::Identifier` too, so I think it's more consistent overall.

That is only a partial implementation because of time constraints, and it could probably have been extended to `UnspecifiedAccountIdentifier` without too much effort, but there's no reason we should only restrict single accounts in the vote manager IMHO, as the public interface we use there is the same for both.

Not a blocker for this PR, but I think we should polish this a bit and uniform the way we access information

Reply:
> That is only a partial implementation because of time constraints

Well, everything is a partial implementation because of time constraints (?). 

I still think it's better to throw an error here and remove it when (if) the feature is fully implemented.

Reply:
> It seems like an optimization to save space in the accounts (so they can be 32 bytes), but I don't know what was Vincent thinking. Multisig accounts seem like a half-baked feature overall.

The wire format can have those optimizations, but after we've validated the witness we could incorporate the info in something like https://github.com/input-output-hk/chain-libs/blob/832f2bbc47f09d95688af8e4c2cef031bddcee74/chain-impl-mockchain/src/transaction/input.rs#L56

Reply:
> The wire format can have those optimizations, but after we've validated the witness we could incorporate the info in something like 

Oh, yeah, sure... To be honest I was just guessing, I'm not sure what's exactly optimized. Inputs are already [tagged](https://github.com/input-output-hk/chain-libs/blob/832f2bbc47f09d95688af8e4c2cef031bddcee74/chain-impl-mockchain/src/transaction/input.rs#L67) , so I'm not sure why not have a tag there.

My guess is that it's because the multisig ledger was supposed to be unified with the normal account ledger or something like that.